client/my-sites/plugins/plugins-list/index.jsx

```diff
@@ -0,0 +1,402 @@
+	},
+
+	setBulkSelectionState( plugins, selectionState ) {
+		let slugsToBeUpdated = {};
+		plugins.forEach( plugin => slugsToBeUpdated[ plugin.slug] = this.hasNoSitesThatCanManage( plugin ) ? false : selectionState );
+
+		this.setState( { selectedPlugins: Object.assign( {}, this.state.selectedPlugins, slugsToBeUpdated ) } );
+	},
+
+	getPluginBySlug( slug ) {
+		for ( let i = 0, l = this.props.plugins.lenght; i < l; i++ ) {
+			if ( this.props.plugins[ i ].slug === slug ) {
+				return this.props.plugins[ i ];
+			}
+		}
+		return false;
+	},
+
+	filterSelection: {
+		active( plugin ) {
+			if ( this.isSelected( plugin ) ) {
+				return plugin.sites.some( site => site.plugin && site.plugin.active );
+			}
+			return false;
+		},
+		inactive( plugin ) {
+			if ( this.isSelected( plugin ) ) {
+				return plugin.sites.some( site => site.plugin && ! site.plugin.active );
+			}
+			return false;
+		},
+		updates( plugin ) {
+			if ( this.isSelected( plugin ) ) {
+				return plugin.sites.some( site => site.plugin && site.plugin.update && site.canUpdateFiles );
+			}
+			return false;
+		},
+		selected( plugin ) {
+			return this.isSelected( plugin );
+		}
+	},
+
+	getPluginUpdateCount() {
+		if ( this.props.isWpCom ) {
+			return 0;
+		}
+		return this.props.pluginUpdateCount;
+	},
+
+	hasNoSitesThatCanManage( plugin ) {
+		return ! plugin.wpcom &&
+			! plugin.sites.some( site => site.modules && site.modules.indexOf( 'manage' ) !== -1 );
+	},
+
+	getSelected() {
+		if ( ! this.props.plugins ) {
+			return [];
+		}
+		return this.props.plugins.filter( this.filterSelection.selected.bind( this ) );
+	},
+
+	siteSuffix() {
+		return ( this.props.selectedSite || this.props.sites.get().length === 1 ) ? '/' + this.props.sites.selected : '';
+	},
+
+	recordEvent( eventAction, includeSelectedPlugins ) {
+		eventAction += ( this.props.selectedSite ? '' : ' on Multisite' );
+		if ( includeSelectedPlugins ) {
+			const pluginSlugs = this.getSelected().map( plugin => plugin.slug );
+			analytics.ga.recordEvent( 'Plugins', eventAction, 'Plugins', pluginSlugs );
+		} else {
+			analytics.ga.recordEvent( 'Plugins', eventAction );
+		}
+	},
+
+	// Actions
+	toggleBulkManagement() {
+		const bulkManagement = ! this.state.bulkManagement;
+
+		this.setState( {
+			bulkManagement
+		} );
+
+		if ( bulkManagement ) {
+			this.recordEvent( 'Clicked Manage' );
+		} else {
+			// Unselect all plugins.
+			this.setState( { selectedPlugins: {} } );
+
+			//PluginsActions.selectPlugins( this.props.sites, 'none' );
+			if ( this.state.notices && ( this.state.notices.completed || this.state.notices.errors ) ) {
+				PluginsActions.removePluginsNotices( this.state.notices.completed.concat( this.state.notices.errors ) );
+			}
+			this.recordEvent( 'Clicked Manage Done' );
+		}
+	},
+
+	doActionOverSelected( actionName, action ) {
+		PluginsActions.removePluginsNotices( this.state.notices.completed.concat( this.state.notices.errors ) );
```

Comment:
Something about this form of list concatenation has always bothered me. It's not that there's anything syntactically wrong about it, but the essence of what's going on seems skewed because of its left bias. We're not really trying to add the errors onto an existing completed list (I think), but rather trying to send the combination of two lists. We can use the list spread operator to convey this _unbiased_ combination:

```js
PluginsActions.removePluginsNotices( [ ...completed, ...errors ] );
```

Of course, the old way of doing things is still possible, but also ugly.

```js
PluginsActions.removePluginsNotices( [].concat( completed ).concat( errors ) );
```